baseplate/clients/memcache/__init__.py

```diff
@@ -129,9 +130,35 @@ class MemcacheContextFactory(ContextFactory):
 
     """
 
-    def __init__(self, pooled_client: PooledClient):
+    PROM_PREFIX = "bp_memcached_pool"
+    PROM_LABELS = ["pool"]
+
+    promTotalConnections = Gauge(
+        f"{PROM_PREFIX}_size",
+        "Maximum size of this pool",
+        PROM_LABELS,
+    )
+
+    promUsedConnections = Gauge(
+        f"{PROM_PREFIX}_in_use",
+        "Number of connections in this pool currently in use",
+        PROM_LABELS,
+    )
+
+    promFreeConnections = Gauge(
+        f"{PROM_PREFIX}_free",
+        "Number of free connections in this pool",
+        PROM_LABELS,
+    )
+
+    def __init__(self, pooled_client: PooledClient, name: str = "memcache"):
         self.pooled_client = pooled_client
 
+        pool = self.pooled_client.client_pool
+        self.promTotalConnections.labels(name).set_function(lambda: pool.max_size)
+        self.promFreeConnections.labels(name).set_function(lambda: len(pool.free))
+        self.promUsedConnections.labels(name).set_function(lambda: len(pool.used))
```

Comment:
why are the metrics set in the init func instead of in the report_memcache_runtime_metrics method?

Reply:
do we need the label "pool=memcache" when the metric name is "bp_memcached_pool"?

Reply:
Yes, that’s what I tried to explain in the caveat section above. If you were to add the metric to the registry when creating the class, that is you call ‘.labels’ there, then that will be shared for each instance of the class regardless of what it connects to.

Reply:
This is seemingly a problem with the statsd metric too.

Reply:
for prom "made on demand" meaning a new instance of the class is initialized?

Reply:
No, when you get `/metrics` the function in `set_function` is called to create the number.